src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -168,16 +152,73 @@ private void FlattenConcatArg(BoundExpression lowered, ArrayBuilder<BoundExpress
                         var rightConstant = boundCoalesce.RightOperand.ConstantValue;
                         if (rightConstant != null && rightConstant.IsString && rightConstant.StringValue.Length == 0)
                         {
-                            flattened.Add(boundCoalesce.LeftOperand);
-                            return;
+                            // If lowered ends in '?? ""', it can never return null.
+                            loweredCanReturnNull = false;
+
+                            // The left operand might be a call to string.Concat(object)
+                            if (boundCoalesce.LeftOperand.Kind == BoundKind.Call &&
+                                TryExtractStringConcatArgsFromBoundCall((BoundCall)boundCoalesce.LeftOperand, out arguments, out _))
+                            {
+                                return true;
+                            }
+
+                            arguments = ImmutableArray<BoundExpression>.Empty.Add(boundCoalesce.LeftOperand);
+                            return true;
                         }
                     }
                     break;
             }
 
-            // fallback - if nothing above worked, leave arg as-is
-            flattened.Add(lowered);
-            return;
+            arguments = default;
+            loweredCanReturnNull = default;
+            return false;
+        }
+
+        private bool TryExtractStringConcatArgsFromBoundCall(BoundCall boundCall, out ImmutableArray<BoundExpression> arguments, out bool callCanReturnNull)
+        {
+            // callCanReturnNull is true if this method returns true, and boundCall is a call to a string.Concat overload which can return null
```

Comment:
Consider putting this comment which seems to describe the "contract" of this method above the signature. Same with `TryExtractStringConcatArgs`.